### PR TITLE
Allow any type with transforms

### LINKED ISSUE #21: Allow transform streams convert output to an array of any objects, instead of a single string
We need to think through the ideal API for this and #20 as well, but something like:

``` js
execa('./cli', [], {
  transforms: {
     // shortcut to a built-in transform we provide
    stdout: 'line',

    stderr: through2(function (chunk, encoding, callback) {
      // custom processing here
    })
  }
}).then(function (result) {
  result.stdout;
  // => an array of lines

  result.stderr;
  // => an array of whatever the custom transform emits
});
```

### PR DESCRIPTION
**Pull request by ehmicky**
Fixes #21. That's our oldest open issue (almost 8 years old)!
This is also a requirement towards fixing #706.

This allows using any type with transforms by using `{transform, objectMode: true}`. The process' return `stdout`/`stderr` will be an array. Typically, this is an array of objects, but this can be any type, including strings (which is very useful for #706).

```js
const transform = async function * (lines) {
	for await (const line of lines) {
		yield JSON.parse(line)
	}
}

const {stdout} = await execa('./jsonlines-output.js', {stdout: {transform, objectMode: true}});
for (const data of stdout) {
	console.log(stdout) // {...}
}
```

This also works for `stdin`. Users can pass objects to `stdin` with `childProcess.stdin.write()`. They can also directly pass an iterable of objects to the `stdin` option.

```js
const transform = async function * (lines) {
	for await (const line of lines) {
		yield JSON.stringify(line)
	}
}

const input = [{event: 'example'}, {event: 'otherExample'}]
await execa('./jsonlines-input.js', {stdin: [input, {transform, objectMode: true}]});
```

While this is a big PR, most of it are code quality related (tests, types, documentation). The runtime part is actually not as big. 

When it comes to the tests, they mostly repeat existing tests, but using object mode instead. I tried to make sure every edge case was tested.

This PR only has runtime changes for users of transforms with `objectMode: true`. I.e. it has no behavior changes for current users.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -18,9 +18,15 @@ type BaseStdioOption =
 	| 'ignore'
 	| 'inherit';
 
-// @todo Use either `Iterable<string>` or `Iterable<Uint8Array>` based on whether `encoding: 'buffer'` is used.
+// @todo Use `string`, `Uint8Array` or `unknown` for both the argument and the return type, based on whether `encoding: 'buffer'` and `objectMode: true` are used.
 // See https://github.com/sindresorhus/execa/issues/694
```
Comment (ehmicky) [inline index.d.ts:22]: I will update that issue accordingly once this PR is merged.

**Comment by sindresorhus** [comment]
Conflict

**Comment by ehmicky** [comment]
Fixed :+1:
